File: app/src/main/java/studio/mandysa/music/ui/fragment/SearchMusicFragment.java
package studio.mandysa.music.ui.fragment;

import android.media.MediaMetadata;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import mandysax.anna2.callback.Callback;
import studio.mandysa.music.service.Media;
import studio.mandysa.music.service.PlayManager;
import studio.mandysa.music.R;
import studio.mandysa.music.logic.Model.MusicModel;
import studio.mandysa.music.logic.Model.NeteaseCloudMusicApi;
import studio.mandysa.music.logic.Model.SearchMusicModel;
import studio.mandysa.music.logic.Network.ServiceCreator;
import studio.mandysa.music.ui.base.BaseLoadingFragment;
import studio.mandysa.music.ui.viewholder.SongViewHolder;
import studio.mandysa.music.ui.viewmodel.SearchViewModel;
import studio.mandysa.music.utils.ImageLoader;
import studio.mandysa.music.utils.TextViewUtils;

public class SearchMusicFragment extends BaseLoadingFragment {

    private final ArrayList<SearchMusicModel> mList = new ArrayList<>();

    private final SongListAdapter mAdapter = new SongListAdapter();

    private SearchViewModel mViewModel;

    private RecyclerView mRvList;

    private int mPage = 1;

    private void init() {

        mViewModel = getViewModel(SearchViewModel.class);

        mRvList = findViewById(R.id.rv_searchMusicAndSinger_view);

        mRvList.setLayoutManager(new LinearLayoutManager(getContext()));
        mRvList.setHasFixedSize(true);
        mRvList.setAdapter(mAdapter);

        nextPage();

        mRvList.addOnScrollListener(new RecyclerView.OnScrollListener() {

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                assert layoutManager != null;
                int lastCompletelyVisibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition();
                if (lastCompletelyVisibleItemPosition == layoutManager.getItemCount() - 1) {
                    nextPage();
                }
            }
        });
    }

    @Override
    public void onRefresh() {
        super.onRefresh();
        nextPage();
    }

    private void nextPage() {
        ServiceCreator.create(NeteaseCloudMusicApi.class).searchMusic(mViewModel.getSearchContent().getValue(), (mPage - 1) * 30).set(new Callback<SearchMusicModel>() {

            @Override
            public void onResponse(boolean loaded, SearchMusicModel t) {
                mList.add(t);
                mAdapter.notifyItemInserted(mList.size());
                if (!loaded) return;
                if (mPage == 1)
                    hide();
                mPage++;
            }

            @Override
            public void onFailure(int code) {
                if (code != 200)
                    show();
            }

        });
    }

    @Override
    protected void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init();
    }

    @Override
    protected void onActivityCreated(Bundle bundle) {
        super.onActivityCreated(bundle);
        mViewModel.getSearchContent().observe(this, p1 -> {
            mPage = 1;
            if (mRvList.getChildCount() > 0) {
                mRvList.removeAllViews();
                mList.clear();
                mAdapter.notifyDataSetChanged();
            }
            onRefresh();
        });
    }

    @Override
    protected int onCreateView() {
        return R.layout.search_music_and_singer;
    }

    public class SongListAdapter extends RecyclerView.Adapter<SongViewHolder> {

        @Override
        public void onBindViewHolder(@NonNull final SongViewHolder p1, final int p2) {
            MediaMetadata mediaMetadata = mList.get(p2).build();
            if (mList.get(p2).picUrl != null)
                ImageLoader.getInstance().load(mList.get(p2).picUrl).into(p1.cover);
            else
                ServiceCreator.create(NeteaseCloudMusicApi.class).getMusicInfo(mediaMetadata.getString(Media.MEDIA_ID)).set(new Callback<MusicModel>() {

                    @Override
                    public void onResponse(boolean loaded, MusicModel t) {
                        ImageLoader.getInstance().load(mList.get(p2).picUrl = t.picUrl).into(p1.cover);
                    }

                    @Override
                    public void onFailure(int code) {
                    }
                });
            p1.songName.setText(mediaMetadata.getString(Media.TITLE));
            p1.singerName.setText(mediaMetadata.getString(Media.ARTIST));
            TextViewUtils.markByColor(p1.songName, mViewModel.getSearchContent().getValue());
            TextViewUtils.markByColor(p1.singerName, mViewModel.getSearchContent().getValue());
            p1.itemView.setOnClickListener(p11 -> {
                ArrayList<MediaMetadata> playList = new ArrayList<>();
                for (SearchMusicModel smm : mList) {
                    playList.add(smm.build());
                }
                PlayManager.getInstance(getActivity()).playMedia(playList, p2);
            });
        }

        @Override
        public int getItemViewType(int position) {
            return position;
        }

        @NonNull
        public SongViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new SongViewHolder(LayoutInflater.from(getContext()).inflate(R.layout.song_item, parent, false));
        }

        public int getItemCount() {
            return mList.size();
        }

    }

}

File: app/src/main/java/studio/mandysa/music/ui/search/searchlist/SearchMusicFragment.kt
package studio.mandysa.music.ui.search.searchlist

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import mandysax.anna2.callback.Callback
import mandysax.lifecycle.livedata.Observer
import mandysax.media.DefaultPlayerManager
import mandysax.media.model.DefaultArtist
import mandysax.media.model.DefaultMusic
import studio.mandysa.jiuwo.utils.addModels
import studio.mandysa.jiuwo.utils.linear
import studio.mandysa.jiuwo.utils.recyclerAdapter
import studio.mandysa.jiuwo.utils.setup
import studio.mandysa.music.R
import studio.mandysa.music.databinding.ItemSongBinding
import studio.mandysa.music.logic.model.MusicModel
import studio.mandysa.music.logic.model.NeteaseCloudMusicApi
import studio.mandysa.music.logic.model.SearchMusicModel
import studio.mandysa.music.logic.network.ServiceCreator
import studio.mandysa.music.logic.utils.create
import studio.mandysa.music.logic.utils.createAlbum
import studio.mandysa.music.logic.utils.markByColor
import studio.mandysa.music.logic.utils.set
import studio.mandysa.music.ui.search.SearchViewModel

class SearchMusicFragment(private val mViewModel: SearchViewModel) : BaseFragment() {

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mViewModel.searchContentLiveData.observe(viewLifecycleOwner) { it ->
            if (!TextUtils.isEmpty(it)) {
                mBinding.let {
                    if (it.recycler.recyclerAdapter.mModel != null) {
                        it.recycler.scrollToPosition(0)
                        it.recycler.recyclerAdapter.clearModels()
                        mPage = 1
                    }
                    it.stateLayout.showLoadingState()
                }
            }
        }
        mBinding.apply {
            stateLayout.apply {
                showLoading {
                    nextPage()
                }
            }
            recycler.linear().setup {
                addType<MusicModel>(R.layout.item_song)
                onBind {
                    ItemSongBinding.bind(itemView).run {
                        val model = getModel<MusicModel>()
                        songCover.setImageURI(model.coverUrl)
                        songName.text = model.title
                        songSingerName.text = model.artist[0].name
                        songName.markByColor(mViewModel.searchContentLiveData.value)
                        songSingerName.markByColor(mViewModel.searchContentLiveData.value)
                        itemView.setOnClickListener {
                            DefaultPlayerManager.getInstance()!!.loadAlbum(
                                models.createAlbum(),
                                modelPosition
                            )
                            DefaultPlayerManager.getInstance()!!.play()
                        }
                        val lifecycleObserver =
                            Observer<DefaultMusic<DefaultArtist>> { p1 ->
                                if (p1.id == model.id) {
                                    songName.setTextColor(
                                        context.getColor(R.color.secondary_color)
                                    )
                                    songSingerName.setTextColor(context.getColor(R.color.secondary_color))
                                } else {
                                    songName.setTextColor(
                                        context.getColor(android.R.color.black)
                                    )
                                    songSingerName.setTextColor(context.getColor(R.color.tv_color_light))
                                    songName.markByColor(mViewModel.searchContentLiveData.value)
                                    songSingerName.markByColor(mViewModel.searchContentLiveData.value)
                                }
                            }
                        onAttached {
                            DefaultPlayerManager.getInstance()!!.changeMusicLiveData()
                                .observe(viewLifecycleOwner, lifecycleObserver)
                        }
                        onDetached {
                            DefaultPlayerManager.getInstance()!!.changeMusicLiveData()
                                .removeObserver(lifecycleObserver)
                        }
                    }
                }
            }
        }
    }

    override fun nextPage() {
        ServiceCreator.create(NeteaseCloudMusicApi::class.java)
            .searchMusic(mViewModel.searchContentLiveData.value, (mPage - 1) * 30)
            .set(viewLifecycleOwner.lifecycle, object : Callback<List<SearchMusicModel>> {
                override fun onResponse(t: List<SearchMusicModel>?) {
                    NeteaseCloudMusicApi::class.java.create()
                        .getMusicInfo(t!!)
                        .set(viewLifecycleOwner.lifecycle, object : Callback<List<MusicModel>> {
                            override fun onResponse(t: List<MusicModel>?) {
                                mBinding.recycler.addModels(t!!)
                                mBinding.stateLayout.showContentState()
                                mPage++
                            }

                            override fun onFailure(code: Int) {
                                mBinding.stateLayout.showErrorState()
                            }
                        })
                }

                override fun onFailure(code: Int) {
                    mBinding.stateLayout.showErrorState()
                }
            })
    }

}